src/pdf.sandbox.js

```diff
@@ -24,86 +25,124 @@ const TESTING =
-      "delete exports;",
-      "delete module;",
-      "delete data;",
+      `pdfjsScripting.initSandbox({ data: ${sandboxData}, external: ${external} })`,
     ];
     if (!TESTING) {
-      code = code.concat(extra.map(name => `delete ${name};`));
       code.push("delete debugMe;");
+      // Don't expose external functions
+      code.push(this._externalNames.map(name => `delete ${name};`).join(";"));
+      delete this._externalNames;
+    }
+    const success = !!this._module.ccall(
+      "initSandbox",
+      "int",
+      ["string", "int"],
+      [code.join("\n"), this._alertOnError]
+    );
+    if (!success) {
+      this.nukeSandbox();
     }
-    this._evalInSandbox(code.join("\n"), this._alertOnError);
-    this._dispatchEventName = data.dispatchEventName;
+    return success;
   }
 
-  dispatchEvent(event) {
-    if (this._dispatchEventName === null) {
-      throw new Error("Sandbox must have been initialized");
+  callFunction(name, args) {
+    if (this._commFun && typeof name === "string") {
+      try {
+        args = Sandbox.cloneIn(args);
+        this._commFun(name, args);
+      } catch (err) {
+        console.error(name, args, err);
+      }
     }
-    event = JSON.stringify(event);
-    this._evalInSandbox(
-      `app["${this._dispatchEventName}"](${event});`,
-      this._alertOnError
-    );
+  }
+
+  dispatchEvent(event) {
+    this.callFunction("dispatchEvent", event);
   }
 
   dumpMemoryUse() {
-    this._module.ccall("dumpMemoryUse", null, []);
+    if (this._module) {
+      this._module.ccall("dumpMemoryUse", null, []);
+    }
   }
 
   nukeSandbox() {
-    this._dispatchEventName = null;
-    this._module.ccall("nukeSandbox", null, []);
-    this._module = null;
-    this._evalInSandbox = null;
+    if (this._module !== null) {
+      this.callFunction("finalize", null);
+      this._module.ccall("nukeSandbox", null, []);
+      this._module = null;
+      this._dispatchEvent = null;
+    }
+  }
+
+  static cloneIn(obj) {
+    // The communication with the Quickjs sandbox is based on strings
+    // So we use JSON.stringfy to serialize
+    return window.JSON.stringify(obj);
+  }
+
+  static cloneOut(obj) {
+    // ...  and JSON.parse to deserialize
+    return window.JSON.parse(obj);
   }
 
   evalForTesting(code, key) {
     if (TESTING) {
-      this._evalInSandbox(
-        `try {
+      this._evalInSandbox = this._module.ccall(
+        "evalInSandbox",
+        null,
+        ["string", "int"],
+        [
+          `try {
            send({ id: "${key}", result: ${code} });
          } catch (error) {
            send({ id: "${key}", result: error.message });
          }`,
-        this._alertOnError
+          this._alertOnError,
+        ]
       );
     }
   }
 }
 
 function QuickJSSandbox() {
+  const handlers = {
+    callFunction(name, args) {},
+    cloneIn: Sandbox.cloneIn,
+    cloneOut: Sandbox.cloneOut,
+  };
+  window.sandboxExternals = buildExternals(window, handlers);
```

Comment:
This works, but I was wondering if there is another way to share values with the module without temporarily polluting the global namespace.